File: compiler/feedback/names.go
package feedback

import (
	"wipple/database"
	"wipple/queries"
	"wipple/visit"
)

func registerNames() {
	register(Feedback[string]{
		Id:    "unresolved",
		Rank:  RankNames,
		Query: queries.Unresolved,
		Render: func(render *Render, node database.Node, name string) {
			render.WriteString("Can't find ")
			render.WriteCode(name)
			render.WriteString(".")
			render.WriteBreak()
			render.WriteString("Double-check your spelling.")
		},
	})

	type ambiguousData struct {
		Name        string
		Definitions []database.Node
	}

	register(Feedback[ambiguousData]{
		Id:   "ambiguous",
		Rank: RankNames,
		Query: func(db *database.Db, node database.Node, filter func(node database.Node) bool, f func(data ambiguousData)) {
			queries.Ambiguous(db, node, filter, func(name string) {
				var definitions []database.Node
				if fact, ok := database.GetFact[visit.ResolvedFact](node); ok {
					for _, definition := range fact.Definitions {
						definitions = append(definitions, definition.GetNode())
					}
				}

				f(ambiguousData{
					Name:        name,
					Definitions: definitions,
				})
			})
		},
		Render: func(render *Render, node database.Node, data ambiguousData) {
			render.WriteCode(data.Name)
			render.WriteString(" could refer to ")

			items := make([]func(), 0, len(data.Definitions))
			for _, definition := range data.Definitions {
				items = append(items, func() {
					render.WriteNode(definition)
				})
			}
			render.WriteList(items, "or", 3)

			render.WriteString(".")
			render.WriteBreak()
			render.WriteString("Rename the extra definitions.")
		},
	})
}

File: compiler/src/feedback/items/names.rs
use crate::{
    feedback::{FeedbackCtx, FeedbackRank, RegisteredFeedback},
    queries,
};
use std::collections::BTreeSet;

pub fn register(ctx: &mut FeedbackCtx) {
    ctx.register(RegisteredFeedback::new(
        "unresolved",
        |ctx, f| {
            queries::unresolved(ctx, &mut |(name, suggestion)| {
                f((ctx.node.clone(), name, suggestion))
            });
        },
        |_| FeedbackRank::Names,
        |(node, _, suggestion)| (node.clone(), BTreeSet::from_iter(suggestion.clone())),
        |w, (_, name, suggestion)| {
            w.write_string("Can't find ");
            w.write_code(name);
            w.write_string(".");
            w.write_break();

            if let Some(suggestion) = suggestion {
                w.write_string("Did you mean ");
                w.write_node(suggestion);
                w.write_string("?");
            } else {
                w.write_string("Double-check your spelling.");
            }
        },
    ));

    ctx.register(RegisteredFeedback::new(
        "ambiguous",
        queries::ambiguous,
        |_| FeedbackRank::Names,
        |(node, _, _)| (node.clone(), BTreeSet::new()),
        |w, (_, name, definitions)| {
            w.write_code(name);
            w.write_string(" could refer to ");

            if !definitions.is_empty() {
                w.write_list("or", 3, |list| {
                    for node in definitions {
                        list.add(|w| w.write_node(node));
                    }
                });
            }

            w.write_string(".");
            w.write_break();
            w.write_string("Rename the extra definitions.");
        },
    ));
}
